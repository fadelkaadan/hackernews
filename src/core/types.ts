export interface IStory {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  text: string;
  time: number;
  title: string;
  type: string;
  url: string;
}

export interface IComment {
  by: string;
  id: number;
  kids: number[];
  parent: number;
  text: string;
  time: number;
  type: string;
}

export interface IUser {
  about: string;
  created: number;
  id: string;
  karma: number;
  submitted: number[];
}