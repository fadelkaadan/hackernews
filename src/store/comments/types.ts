export interface IComment {
  by: string;
  id: number;
  kids: number[];
  parent: number;
  text: string;
  time: number;
  type: string;
}

export interface CommentsState {
  list: IComment[];
  pending: Boolean;
  error: Object | null;
}

export const FETCH_COMMENTS_SUCCESS = "FETCH_COMMENTS_SUCCESS";
export const FETCH_COMMENTS_PENDING = "FETCH_COMMENTS_PENDING";
export const FETCH_COMMENTS_ERROR = "FETCH_COMMENTS_ERROR";

interface IFetchCommentsPendingAction {
  type: string;
}

interface IFetchCommentsSuccessAction {
  type: string;
  data: IComment[];
}

interface IFetchCommentsErrorAction {
  type: string;
  error: any;
}

export type CommentsActionTypes =
  | IFetchCommentsPendingAction
  | IFetchCommentsSuccessAction
  | IFetchCommentsErrorAction;
