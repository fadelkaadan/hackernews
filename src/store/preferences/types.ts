export enum ThemeEnum {
  DARK = "DARK",
  NIGHT = "NIGHT",
  LIGHT = "LIGHT",
}

export interface ThemeState {
  isPreferencesClicked: Boolean
  theme: ThemeEnum;
  isOneColumnLayout: Boolean;
}

export const TOGGLE_PREFERENCES = "TOGGLE_PREFERENCES";
export const SET_THEME = "SET_THEME";
export const TOGGLE_ONE_COLUMN_LAYOUT = "TOGGLE_ONE_COLUMN_LAYOUT";

interface togglePreferences {
  type: typeof TOGGLE_PREFERENCES;
}

interface setThemeAction {
  type: typeof SET_THEME;
  data: ThemeEnum;
}

interface toggleOneColumnLayoutAction {
  type: typeof TOGGLE_ONE_COLUMN_LAYOUT;
}

export type ThemeActionTypes =
  | setThemeAction
  | toggleOneColumnLayoutAction
  | togglePreferences;
