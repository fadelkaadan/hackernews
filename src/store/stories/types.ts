export interface IStory {
  by: string;
  descendants: number;
  id: number;
  kids: number[];
  score: number;
  text: string;
  time: number;
  title: string;
  type: string;
  url: string;
}

export interface StoriesState {
  list: IStory[] | undefined;
  pending: Boolean;
  error: Object | null;
}

export const FETCH_STORIES_SUCCESS = "FETCH_STORIES_SUCCESS";
export const FETCH_STORIES_PENDING = "FETCH_STORIES_PENDING";
export const FETCH_STORIES_ERROR = "FETCH_STORIES_ERROR";

interface fetchStoriesAction {
  type: string;
  data?: IStory[];
  error?: any;
}

export type StoriesActionTypes = fetchStoriesAction;
