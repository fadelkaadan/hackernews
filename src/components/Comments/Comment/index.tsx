import React from "react";

interface IComment {
  by: string;
  id: number;
  kids: number[];
  parent: number;
  text: string;
  time: number;
  type: string;
}

interface CommentProps {
  author: string;
  subcomments: IComment[];
  text: string;
  time: number;
}

const Comment = ({ author, subcomments, text, time }: CommentProps) => {
  return <div>Comment</div>;
};

export default Comment;
