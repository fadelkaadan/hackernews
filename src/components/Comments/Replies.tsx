import React, { useState } from "react";
import CommentsContainer from "../../containers/CommentsContainer";
import CommentsButton from "./CommentsButton";
import { faReply } from "@fortawesome/free-solid-svg-icons";

interface RepliesProps {
  data: number[];
}

const Replies = ({ data }: RepliesProps) => {
  const [isCommentsOpen, setIsCommentsOpen] = useState(false);

  const handleClick = () => {
    setIsCommentsOpen(!isCommentsOpen);
  };

  return (
    <>
      <CommentsButton
        comments={data}
        isOpen={isCommentsOpen}
        onClick={handleClick}
        label={data && data.length === 1 ? "reply" : "replies"}
        icon={faReply}
      />
      {isCommentsOpen ? <CommentsContainer commentIds={data} /> : null}
    </>
  );
};

export default Replies;
