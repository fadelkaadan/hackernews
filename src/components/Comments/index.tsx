import React from "react";
import styled from "styled-components";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faComment } from "@fortawesome/free-regular-svg-icons";

interface CommentsProps {
  comments: number[];
}

const Wrapper = styled.div`
  font-size: 12px;
  margin-left: 10px;
`

const Comments = ({ comments }: CommentsProps) => {
  const getCommentsCount = () => {
    if (comments) {
      return comments.length
    }
    return 0;
  }

  return (
    <Wrapper>
      <p><FontAwesomeIcon icon={faComment}/> {getCommentsCount()}</p>
    </Wrapper>
  );
};

export default Comments;
