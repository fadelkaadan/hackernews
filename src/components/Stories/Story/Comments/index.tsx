import React from "react";
import styled from "styled-components";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faComment } from "@fortawesome/free-regular-svg-icons";

interface CommentsProps {
  comments: number[];
}

const Wrapper = styled.p`
  font-size: 1.2rem;
  color: ${(props) => props.theme.main.text.secondary};
  border: 0.1rem solid ${(props) => props.theme.main.text.secondary};
  border-radius: 0.5rem;
  padding: 0.3rem 0.5rem;
  cursor: pointer;

  &:hover {
    color: ${(props) => props.theme.main.text.primary};
    border: 0.1rem solid ${(props) => props.theme.main.text.primary};
  }
`;

const WrapperDisabled = styled(Wrapper)`
  color: ${(props) => props.theme.main.text.disabled};
  border: 0.1rem solid ${(props) => props.theme.main.text.disabled};

  &:hover {
    color: ${(props) => props.theme.main.text.disabled};
    border: 0.1rem solid ${(props) => props.theme.main.text.disabled};
    cursor: default;
  }
`;

const Comments = ({ comments }: CommentsProps) => {
  const getCommentsCount = () => {
    if (comments) return comments.length;
    return 0;
  };

  return getCommentsCount() === 0 ? (
    <WrapperDisabled>
      <FontAwesomeIcon icon={faComment} /> {getCommentsCount()} comments
    </WrapperDisabled>
  ) : (
    <Wrapper>
      <FontAwesomeIcon icon={faComment} /> {getCommentsCount()} comments
    </Wrapper>
  );
};

export default Comments;
