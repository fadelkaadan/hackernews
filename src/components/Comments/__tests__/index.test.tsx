import React from "react";
import { render, screen, cleanup, RenderResult } from "../../../utilities/test-utils";
import Comments from "../";
import { IComment } from "../../../core/types";

describe("Comments", () => {
  afterEach(cleanup);

  const fakeContent = "Here is some content";
  const fakeContent2 = "Here is some content 2";

  const comments: IComment[] = [
    {
      by: "fadel",
      id: 1,
      kids: [1, 2, 3],
      parent: 12,
      text: fakeContent,
      time: 1,
      type: "comment",
    },
    {
      by: "adel",
      id: 2,
      kids: [4, 5, 6],
      parent: 22,
      text: fakeContent2,
      time: 2,
      type: "comment",
    },
  ];

  it("Should match the snapshot", () => {
    const { container }: RenderResult = render(
      <Comments data={comments} isPending={false} errorMessage={{}} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should render Comments", () => {
    render(<Comments data={comments} isPending={false} errorMessage={{}} />);
    const authors: HTMLElement[] = screen.getAllByRole("link");
    expect(authors.length).toEqual(2);
    const content: HTMLElement = screen.getByText(fakeContent);
    expect(content.textContent).toEqual(fakeContent);
    const content2: HTMLElement = screen.getByText(fakeContent2);
    expect(content2.textContent).toEqual(fakeContent2);
    const buttons: HTMLElement[] = screen.getAllByRole("button");
    expect(buttons.length).toEqual(2);
  });

  it("Should render a loading indicator when content is loading", () => {
    const { container } = render(<Comments data={comments} isPending={true} errorMessage={{}} />);
    const authors: HTMLElement[] = screen.queryAllByRole("link");
    expect(authors.length).toBe(0);
    const loadingIndicator: SVGSVGElement | null = container.querySelector("svg");
    expect(loadingIndicator).toBeTruthy();
  });
});
