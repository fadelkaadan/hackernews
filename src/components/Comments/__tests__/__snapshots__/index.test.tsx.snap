// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments Should match the snapshot 1`] = `
<div
  class="sc-jrsJCI dmJLbr"
>
  <div
    class="sc-fujyUd eDtkpJ"
  >
    <div
      class="sc-jSFkmK EqSxB"
    >
      <div
        class="sc-gtssRu UymyY"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user-circle fa-w-16 "
          data-icon="user-circle"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M248 104c-53 0-96 43-96 96s43 96 96 96 96-43 96-96-43-96-96-96zm0 144c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm0-240C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-49.7 0-95.1-18.3-130.1-48.4 14.9-23 40.4-38.6 69.6-39.5 20.8 6.4 40.6 9.6 60.5 9.6s39.7-3.1 60.5-9.6c29.2 1 54.7 16.5 69.6 39.5-35 30.1-80.4 48.4-130.1 48.4zm162.7-84.1c-24.4-31.4-62.1-51.9-105.1-51.9-10.2 0-26 9.6-57.6 9.6-31.5 0-47.4-9.6-57.6-9.6-42.9 0-80.6 20.5-105.1 51.9C61.9 339.2 48 299.2 48 256c0-110.3 89.7-200 200-200s200 89.7 200 200c0 43.2-13.9 83.2-37.3 115.9z"
            fill="currentColor"
          />
        </svg>
        <a
          class="sc-dlnjPT jAMKpE"
          href="/users/fadel"
        >
          fadel
        </a>
      </div>
      <p
        class="sc-eCApGN EpjZS"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-clock fa-w-16 "
          data-icon="clock"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm61.8-104.4l-84.9-61.7c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v141.7l66.8 48.6c5.4 3.9 6.5 11.4 2.6 16.8L334.6 349c-3.9 5.3-11.4 6.5-16.8 2.6z"
            fill="currentColor"
          />
        </svg>
         
        51 years ago
      </p>
    </div>
    <div
      class="sc-bdnylx iPnIIX"
    >
      Here is some content
    </div>
    <button
      class="sc-gKAblj bkCUog"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-reply fa-w-16 "
        data-icon="reply"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.309 189.836L184.313 37.851C199.719 24.546 224 35.347 224 56.015v80.053c160.629 1.839 288 34.032 288 186.258 0 61.441-39.581 122.309-83.333 154.132-13.653 9.931-33.111-2.533-28.077-18.631 45.344-145.012-21.507-183.51-176.59-185.742V360c0 20.7-24.3 31.453-39.687 18.164l-176.004-152c-11.071-9.562-11.086-26.753 0-36.328z"
          fill="currentColor"
        />
      </svg>
       
      3 replies
    </button>
  </div>
  <div
    class="sc-fujyUd eDtkpJ"
  >
    <div
      class="sc-jSFkmK EqSxB"
    >
      <div
        class="sc-gtssRu UymyY"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-user-circle fa-w-16 "
          data-icon="user-circle"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M248 104c-53 0-96 43-96 96s43 96 96 96 96-43 96-96-43-96-96-96zm0 144c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm0-240C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-49.7 0-95.1-18.3-130.1-48.4 14.9-23 40.4-38.6 69.6-39.5 20.8 6.4 40.6 9.6 60.5 9.6s39.7-3.1 60.5-9.6c29.2 1 54.7 16.5 69.6 39.5-35 30.1-80.4 48.4-130.1 48.4zm162.7-84.1c-24.4-31.4-62.1-51.9-105.1-51.9-10.2 0-26 9.6-57.6 9.6-31.5 0-47.4-9.6-57.6-9.6-42.9 0-80.6 20.5-105.1 51.9C61.9 339.2 48 299.2 48 256c0-110.3 89.7-200 200-200s200 89.7 200 200c0 43.2-13.9 83.2-37.3 115.9z"
            fill="currentColor"
          />
        </svg>
        <a
          class="sc-dlnjPT jAMKpE"
          href="/users/adel"
        >
          adel
        </a>
      </div>
      <p
        class="sc-eCApGN EpjZS"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-clock fa-w-16 "
          data-icon="clock"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm61.8-104.4l-84.9-61.7c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v141.7l66.8 48.6c5.4 3.9 6.5 11.4 2.6 16.8L334.6 349c-3.9 5.3-11.4 6.5-16.8 2.6z"
            fill="currentColor"
          />
        </svg>
         
        51 years ago
      </p>
    </div>
    <div
      class="sc-bdnylx iPnIIX"
    >
      Here is some content 2
    </div>
    <button
      class="sc-gKAblj bkCUog"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-reply fa-w-16 "
        data-icon="reply"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.309 189.836L184.313 37.851C199.719 24.546 224 35.347 224 56.015v80.053c160.629 1.839 288 34.032 288 186.258 0 61.441-39.581 122.309-83.333 154.132-13.653 9.931-33.111-2.533-28.077-18.631 45.344-145.012-21.507-183.51-176.59-185.742V360c0 20.7-24.3 31.453-39.687 18.164l-176.004-152c-11.071-9.562-11.086-26.753 0-36.328z"
          fill="currentColor"
        />
      </svg>
       
      3 replies
    </button>
  </div>
</div>
`;
